File: cpp/rocketmq-client-cpp-tests/cpp4.x/test/offset/OffsetTest.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <chrono>
#include <iostream>
#include <cassert>

#include <gtest/gtest.h>
#include <spdlog/logger.h>
#include <string>
#include <rocketmq/MQMessageExt.h>
#include <rocketmq/DefaultMQPushConsumer.h>
#include <rocketmq/MQMessageListener.h>

#include "enums/MessageType.h"
#include "frame/BaseOperate.h"
#include "resource/Resource.h"
#include "utils/NameUtils.h"
#include "utils/RandomUtils.h"
#include "utils/VerifyUtils.h"
#include "utils/SimpleConcurrentHashMapUtils.h"
#include "utils/SimpleConcurrentVectorUtils.h"
#include "factory/ConsumerFactory.h"
#include "factory/ProducerFactory.h"
#include "factory/MessageFactory.h"

extern std::shared_ptr<spdlog::logger> multi_logger;
extern std::shared_ptr<Resource> resource;

//Send 10 messages, set other groupid's pushconsumer consumption from first, expect to accept all messages again
TEST(OffsetTest, testConsumeFromFisrtOffset){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::NORMAL, "testConsumeFromFisrtOffset", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group1 = getGroupId("testConsumeFromFisrtOffset1");
    std::string group2 = getGroupId("testConsumeFromFisrtOffset2");
    std::string tag = NameUtils::getRandomTagName();

    std::shared_ptr<RMQNormalListener> listener1 = std::make_shared<RMQNormalListener>("Listener1");
    std::shared_ptr<RMQNormalListener> listener2 = std::make_shared<RMQNormalListener>("Listener2");

    auto rmqPushConsumer1 = std::make_shared<rocketmq::DefaultMQPushConsumer>(group1);
    rmqPushConsumer1->setNamesrvAddr(resource->getNamesrv());
    rmqPushConsumer1->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
    rmqPushConsumer1->setConsumeFromWhere(rocketmq::CONSUME_FROM_LAST_OFFSET);
    rmqPushConsumer1->setConsumeThreadCount(4);
    rmqPushConsumer1->subscribe(topic, tag);
    rmqPushConsumer1->registerMessageListener(listener1.get());
    rmqPushConsumer1->start();

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group1);
    std::this_thread::sleep_for(std::chrono::seconds(2));
    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));
    pullConsumer->shutdown();

    auto producer = ProducerFactory::getRMQProducer(group1);

    ASSERT_NE(producer, nullptr);

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
        producer->send(message);
    }

    long endTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+240*1000L;
    while(endTime > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        if(listener1->getDequeueMessages()->getDataSize() == SEND_NUM){
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }

    ASSERT_EQ(listener1->getDequeueMessages()->getDataSize(),SEND_NUM);

    rmqPushConsumer1->shutdown();

    multi_logger->info("first pushconsumer end");

    auto rmqPushConsumer2 = std::make_shared<rocketmq::DefaultMQPushConsumer>(group2);
    rmqPushConsumer2->setNamesrvAddr(resource->getNamesrv());
    rmqPushConsumer2->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
    rmqPushConsumer2->setConsumeFromWhere(rocketmq::CONSUME_FROM_FIRST_OFFSET);
    rmqPushConsumer2->setConsumeThreadCount(4);
    rmqPushConsumer2->subscribe(topic, tag);
    rmqPushConsumer2->registerMessageListener(listener2.get());
    rmqPushConsumer2->start();

    long endTime2 = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+240*1000L;
    while(endTime2 > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        if(listener2->getDequeueMessages()->getDataSize() == SEND_NUM){
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }

    ASSERT_EQ(listener2->getDequeueMessages()->getDataSize(),SEND_NUM);


    rmqPushConsumer2->shutdown();
    producer->shutdown();
}

//send 10 messages, PullConsumer normally receives messages, but does not update messages offset, expect the messages are receive again
TEST(OffsetTest, test_pull_receive_nack){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::NORMAL, "test_pull_receive_nack", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("test_pull_receive_nack");
    std::string tag = NameUtils::getRandomTagName();

    auto pullConsumer = ConsumerFactory::getPullConsumer(topic,group);
    std::this_thread::sleep_for(std::chrono::seconds(2));
    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer));

    auto producer = ProducerFactory::getRMQProducer(group);
    ASSERT_NE(producer, nullptr);

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,tag+"-"+std::to_string(i));
        producer->send(message);
    }
    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_EQ(SEND_NUM,producer->getEnqueueMessages()->getDataSize());

    std::vector<rocketmq::MQMessageQueue> mqs;
    pullConsumer->fetchSubscribeMessageQueues(topic, mqs);

    std::vector<rocketmq::MQMessageExt> receivedMessage;
    SimpleConcurrentHashMap<std::string,std::atomic<int>> map;

    long endTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+30* 1000L;
    while(endTime > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        std::vector<std::function<void()>> runnables;
        for (auto& mq : mqs) {
            runnables.push_back([&](){
                long long offset = pullConsumer->fetchConsumeOffset(mq, false);
                if(offset<0) return;
                rocketmq::PullResult pullResult = pullConsumer->pull(mq, tag, offset, SEND_NUM);
                switch (pullResult.pullStatus) {
                    case rocketmq::FOUND:
                        for(int j=0;j<pullResult.msgFoundList.size();j++){
                            std::string msgId = pullResult.msgFoundList[j].getMsgId();
                            if(map.contains(msgId)){
                                map[msgId]++;
                            }else{
                                std::atomic<int> val(1);
                                map.insert(msgId,val.load());
                            }
                        }
                        break;
                    case rocketmq::NO_MATCHED_MSG:
                        break;
                    case rocketmq::NO_NEW_MSG:
                        break;
                    case rocketmq::OFFSET_ILLEGAL:
                        break;
                    default:
                        break;
                }
            });
        }
        
        std::vector<std::future<void>> futures;
        for (const auto& runnable : runnables) {
            futures.push_back(std::async(std::launch::async, runnable));
        }

        for (auto& future : futures) {
            future.get();
        }
    }

    for(auto& value : map.getAllValues()){
        ASSERT_TRUE(value>1);
    }

    pullConsumer->shutdown();
    producer->shutdown();
}

File: java/e2e-v4/src/test/java/org/apache/rocketmq/offset/OffsetTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.offset;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.Map.Entry;
import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.apache.rocketmq.client.consumer.DefaultMQPushConsumer;
import org.apache.rocketmq.client.consumer.MessageSelector;
import org.apache.rocketmq.client.consumer.PullResult;
import org.apache.rocketmq.client.consumer.listener.ConsumeConcurrentlyContext;
import org.apache.rocketmq.client.consumer.listener.ConsumeConcurrentlyStatus;
import org.apache.rocketmq.client.consumer.listener.MessageListenerConcurrently;
import org.apache.rocketmq.client.consumer.rebalance.AllocateMessageQueueAveragely;
import org.apache.rocketmq.client.exception.MQBrokerException;
import org.apache.rocketmq.client.exception.MQClientException;
import org.apache.rocketmq.client.rmq.RMQNormalConsumer;
import org.apache.rocketmq.client.rmq.RMQNormalProducer;
import org.apache.rocketmq.common.consumer.ConsumeFromWhere;
import org.apache.rocketmq.common.message.MessageExt;
import org.apache.rocketmq.common.message.MessageQueue;
import org.apache.rocketmq.common.protocol.heartbeat.MessageModel;
import org.apache.rocketmq.enums.TESTSET;
import org.apache.rocketmq.factory.ConsumerFactory;
import org.apache.rocketmq.factory.ProducerFactory;
import org.apache.rocketmq.frame.BaseOperate;
import org.apache.rocketmq.listener.rmq.concurrent.RMQNormalListener;
import org.apache.rocketmq.remoting.exception.RemotingException;
import org.apache.rocketmq.server.batch.BatchProducerTest;
import org.apache.rocketmq.utils.NameUtils;
import org.apache.rocketmq.utils.RandomUtils;
import org.apache.rocketmq.utils.TestUtils;
import org.apache.rocketmq.utils.VerifyUtils;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@Tag(TESTSET.OFFSET)
public class OffsetTest extends BaseOperate {
    private final Logger log = LoggerFactory.getLogger(OffsetTest.class);
    private String tag;
    private final static int SEND_NUM = 10;

    @BeforeEach
    public void setUp() {
        tag = NameUtils.getRandomTagName();
    }

    @Test
    @DisplayName("Send 10 messages, set other groupid's pushconsumer consumption from first, expect to accept all messages again")
    public void testConsumeFromFisrtOffset() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(methodName);
        String groupId1 = getGroupId(methodName);
        String groupId2 = getGroupId(methodName);
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);

        RMQNormalConsumer pullConsumer = ConsumerFactory.getRMQLitePullConsumer(namesrvAddr, groupId1, rpcHook, 1);
        pullConsumer.subscribeAndStartLitePull(topic, MessageSelector.byTag(tag));
        VerifyUtils.tryReceiveOnce(pullConsumer.getLitePullConsumer());
        pullConsumer.shutdown();

        DefaultMQPushConsumer pushConsumer;
        ConcurrentLinkedDeque<MessageExt> deque = new ConcurrentLinkedDeque<>();
        try {
            pushConsumer = new DefaultMQPushConsumer(groupId1, rpcHook, new AllocateMessageQueueAveragely());
            pushConsumer.setInstanceName(RandomUtils.getStringByUUID());
            pushConsumer.setNamesrvAddr(namesrvAddr);
            pushConsumer.subscribe(topic, tag);
            pushConsumer.setConsumeFromWhere(ConsumeFromWhere.CONSUME_FROM_LAST_OFFSET);
            pushConsumer.setMessageListener(new MessageListenerConcurrently() {
                @Override
                public ConsumeConcurrentlyStatus consumeMessage(List<MessageExt> msgs,
                        ConsumeConcurrentlyContext context) {
                    for (MessageExt message : msgs) {
                        log.info("receive message:{}", message);
                        deque.add(message);
                    }
                    return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
                }
            });
            pushConsumer.start();
        } catch (MQClientException e) {
            throw new RuntimeException(e);
        }

        Assertions.assertNotNull(producer);

        producer.send(topic, tag, SEND_NUM);

        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");

        await().atMost(30, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return deque.size() == SEND_NUM;
            }
        });

        pushConsumer.shutdown();
        deque.clear();

        DefaultMQPushConsumer reConsumer;
        try {
            reConsumer = new DefaultMQPushConsumer(groupId2, rpcHook, new AllocateMessageQueueAveragely());
            reConsumer.setInstanceName(RandomUtils.getStringByUUID());
            reConsumer.setNamesrvAddr(namesrvAddr);
            reConsumer.subscribe(topic, tag);
            reConsumer.setConsumeFromWhere(ConsumeFromWhere.CONSUME_FROM_FIRST_OFFSET);
            reConsumer.setMessageListener(new MessageListenerConcurrently() {
                @Override
                public ConsumeConcurrentlyStatus consumeMessage(List<MessageExt> msgs,
                        ConsumeConcurrentlyContext context) {
                    for (MessageExt message : msgs) {
                        log.info("reconsumer received message:{}", message);
                        deque.add(message);
                    }
                    return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
                }
            });
            reConsumer.start();
        } catch (MQClientException e) {
            throw new RuntimeException(e);
        }

        TestUtils.waitForSeconds(30);

        Assertions.assertEquals(SEND_NUM, deque.size(), "reconsumer receive message failed");

        reConsumer.shutdown();
        producer.shutdown();

    }

    @Test
    @DisplayName("Backlog 100 messages, start the consumer, and set the pull message from the LAST, expect to consume 100 messages")
    public void testConsumeFromLastOffset() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(methodName);
        String groupId = getGroupId(methodName);
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);

        Assertions.assertNotNull(producer);

        producer.send(topic, tag, 100);

        Assertions.assertEquals(100, producer.getEnqueueMessages().getDataSize(), "send message failed");

        DefaultMQPushConsumer pushConsumer;
        ConcurrentLinkedDeque<MessageExt> deque = new ConcurrentLinkedDeque<>();
        try {
            pushConsumer = new DefaultMQPushConsumer(groupId, rpcHook, new AllocateMessageQueueAveragely());
            pushConsumer.setInstanceName(RandomUtils.getStringByUUID());
            pushConsumer.setNamesrvAddr(namesrvAddr);
            pushConsumer.subscribe(topic, tag);
            pushConsumer.setConsumeFromWhere(ConsumeFromWhere.CONSUME_FROM_LAST_OFFSET);
            pushConsumer.setMessageListener(new MessageListenerConcurrently() {
                @Override
                public ConsumeConcurrentlyStatus consumeMessage(List<MessageExt> msgs,
                        ConsumeConcurrentlyContext context) {
                    for (MessageExt message : msgs) {
                        log.info("receive message:{}", message);
                        deque.add(message);
                    }
                    return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
                }
            });
            pushConsumer.start();
        } catch (MQClientException e) {
            throw new RuntimeException(e);
        }

        TestUtils.waitForSeconds(30);
        Assertions.assertEquals(100, deque.size(), "consumer receive message failed");

        pushConsumer.shutdown();
        producer.shutdown();

    }

    @Test
    @DisplayName("send 10 messages, PullConsumer normally receives messages, but does not update messages offset, expect the messages are receive again")
    public void test_pull_receive_nack() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(methodName);
        String groupId = getGroupId(methodName);

        RMQNormalConsumer consumer = ConsumerFactory.getRMQPullConsumer(namesrvAddr, groupId, rpcHook);
        consumer.startDefaultPull();
        VerifyUtils.tryReceiveOnce(consumer.getPullConsumer(), topic, tag, 32);
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);
        Assertions.assertNotNull(producer, "Get producer failed");

        producer.send(topic, tag, SEND_NUM);

        TestUtils.waitForSeconds(1);

        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");

        Set<MessageQueue> receiveMessageQueues = null;
        try {
            receiveMessageQueues = consumer.getPullConsumer().fetchSubscribeMessageQueues(topic);
        } catch (MQClientException e) {
            Assertions.fail("Fail to fetchSubscribeMessageQueues");
        }

        Collection<MessageExt> sendCollection = Collections
                .synchronizedCollection(producer.getEnqueueMessages().getAllData());
        Set<MessageQueue> finalMessageQueues = receiveMessageQueues;
        CompletableFuture[] futures = new CompletableFuture[receiveMessageQueues.size()];
        ConcurrentHashMap<String, AtomicInteger> messageMap = new ConcurrentHashMap<>();
        int mqCount = 0;
        for (MessageQueue mq : finalMessageQueues) {
            CompletableFuture<Void> future = CompletableFuture.supplyAsync(() -> {
                try {
                    long offset = consumer.getPullConsumer().fetchConsumeOffset(mq, false);
                    if (offset < 0)
                        return null;
                    long startTime = System.currentTimeMillis();
                    while (System.currentTimeMillis() < startTime + 5000) {
                        PullResult pullResult = consumer.getPullConsumer().pull(mq, tag, offset, SEND_NUM);
                        switch (pullResult.getPullStatus()) {
                            case FOUND:
                                List<MessageExt> messages = pullResult.getMsgFoundList();
                                for (MessageExt message : messages) {
                                    log.info("MessageId:{}, Body:{}, Property:{}, Retry:{}", message.getMsgId(),
                                            StandardCharsets.UTF_8.decode(ByteBuffer.wrap(message.getBody())),
                                            message.getProperties(), message.getReconsumeTimes());
                                    sendCollection
                                            .removeIf(messageExt -> messageExt.getMsgId().equals(message.getMsgId()));
                                    if (messageMap.containsKey(message.getMsgId())) {
                                        messageMap.get(message.getMsgId()).incrementAndGet();
                                    } else {
                                        messageMap.put(message.getMsgId(), new AtomicInteger(1));
                                    }
                                }
                                break;
                            case NO_MATCHED_MSG:
                                break;
                            case NO_NEW_MSG:
                                break;
                            case OFFSET_ILLEGAL:
                                break;
                            default:
                                break;
                        }
                    }
                } catch (MQBrokerException e) {
                    e.printStackTrace();
                    Assertions.fail("Pull fetch message error");
                } catch (RemotingException e) {
                    e.printStackTrace();
                    Assertions.fail("Pull fetch message error");
                } catch (InterruptedException e) {
                    e.printStackTrace();
                    Assertions.fail("Pull fetch message error");
                } catch (MQClientException e) {
                    e.printStackTrace();
                    Assertions.fail("Pull fetch message error");
                }
                return null;
            });
            futures[mqCount++] = future;
        }
        try {
            CompletableFuture.allOf(futures).get(60, TimeUnit.SECONDS);
        } catch (Exception e) {
            e.printStackTrace();
            Assertions.fail("receive response count not match");
        }
        log.info("A total of {} messages were received", messageMap.size());
        for (Entry<String, AtomicInteger> entry : messageMap.entrySet()) {
            Assertions.assertTrue(entry.getValue().get() >= 1, "message receive count not match");
        }
    }
}
